nixos/modules/virtualisation/qemu-vm.nix

```diff
@@ -478,7 +493,27 @@ in
 
   config = {
 
-    boot.loader.grub.device = mkVMOverride cfg.bootDevice;
+    # Note [Disk layout with `useBootLoader`]
+    #
+    # If `useBootLoader = true`, we configure 2 drives:
+    # `/dev/?da` for the root disk, and `/dev/?db` for the boot disk
+    # which has the `/boot` partition and the boot loader.
+    # Concretely:
+    #
+    # * The second drive's image `disk.img` is created in `bootDisk = ...`
+    #   using a throwaway VM. Note that there the disk is always `/dev/vda`,
+    #   even though in the final VM it will be at `/dev/*b`.
+    # * The disks are attached in `virtualisation.qemu.drives`.
+    #   Their order makes them appear as devices `a`, `b`, etc.
+    # * `fileSystems."/boot"` is adjusted to be on device `b`.
+
+    # If `useBootLoader`, GRUB goes to the second disk, see
+    # note [Disk layout with `useBootLoader`].
+    boot.loader.grub.device = mkVMOverride (
+      if cfg.useBootLoader
+        then driveDeviceName 1 # second disk
```

Comment:
Oh, this comment makes me sad.  Seems like driveDeviceName should just use 1-based indices instead.

Reply:
I changed #92205 to use 1-based indices.

Reply:
@chkno Personally I think 0-based indices are also fine; I'd put this comment in anyway, to align with the `GRUB goes to the second disk` comment further up, so that it's really clear that this is where that's done.

Reply:
Updated by rebasing and switching to `driveDeviceName 2 # second disk`.